packages/kas/src/parser.ts

```diff
@@ -0,0 +1,387 @@
+                            left = Mul.fold(Mul.handleDivide(left, right)); // left associative
+                            break;
+                    }
+                } else {
+                    break;
+                }
+            }
+        }
+
+        return left;
+    }
+
+    negative() {
+        if (this.peek().kind === "-") {
+            this.consume(); // -
+            return Mul.handleNegative(this.negative());
+        } else {
+            return this.triglog();
+        }
+    }
+
+    trig(): [string] {
+        const next = this.peek();
+        if (next.kind === "TRIG") {
+            this.consume(); // trig
+            return [next.value];
+        } else {
+            throw new Error(`Expected TRIG token but tog ${next.kind}`);
+        }
+    }
+
+    trigfunc(): [string] | [string, Expr] {
+        const index = this.index; // save state
+        try {
+            const left = this.trig();
+            const op = this.peek();
+            if (op.kind === "^") {
+                this.consume(); // ^
+                const right = this.negative();
+                return [left[0], right];
+            } else {
+                return left;
+            }
+        } catch (e) {
+            this.index = index; // restore state
+            const next = this.peek();
+            if (next.kind === "TRIGINV") {
+                this.consume(); // triginv
+                return [next.value];
+            } else {
+                throw new Error(`Expected TRIGINV but got ${next.kind}`);
+            }
+        }
+    }
+
+    logbase() {
+        const next = this.peek();
+        this.consume();
+        switch (next.kind) {
+            case "ln":
+                return Log.natural();
+            case "log":
+                if (this.peek().kind === "_") {
+                    this.consume();
+                    return this.subscriptable();
+                } else {
+                    return Log.common();
+                }
+            default:
+                throw new Error(
+                    `Expected either ln or log but got ${next.kind}`,
+                );
+        }
+    }
+
+    triglog() {
+        const index = this.index; // save state
+        try {
+            const left = this.trigfunc();
+            const right = this.negative();
+            return Trig.create(left, right);
+        } catch {
+            this.index = index; // restore state
+            try {
+                const left = this.logbase();
+                const right = this.negative();
+                return Log.create(left, right);
+            } catch {
+                this.index = index; // restore state
+                return this.power();
+            }
+        }
+    }
+
+    power() {
+        const left = this.primitive();
+        const op = this.peek();
+        if (op.kind === "^") {
+            this.consume(); // ^
+            const right = this.negative();
+            return new Pow(left, right);
```

Comment:
TODO: add a comment here about this being right associative and why that's the case.